### DISCUSSION TITLE
Improvement to debugger addon

### DISCUSSION THREAD

**Discussion by sfederici** [ScratchAddons/ScratchAddons#7267] [Ideas]
When I watch my students' story-telling projects, I spot often points where I noticed something wrong or something that could be improved, but, in order to check if this is true, I have to

- quickly pause the project
- restore hidden sprites that were visible just a moment before
- hide sprites that just showed up
- restore the right backgrounds
- cut the scripts in order to rerun them from the relevant block(s).

All these steps (that could be useless as I could just perform them in the wrong way, as I don't know the project yet) could be overcome if in the debugger, other than the "step forward" button it would be available a "step backward" button that brings me back to the previous situation from where I can start the project again.

I guess this can be difficult -or even very difficult- but maybe not, so...

**Comment by WorldLanguages** [comment]
I've seen "Scratch time travel debugging" PDFs on the internet, but as far as I've seen they only got stuck as published papers and not actual useful public software...

**Reply by WorldLanguages** [reply]
Nevermind, see https://scratch.ugent.be/blink/editor/

**Comment by sfederici** [comment]
Excellent! It really works https://scratch.ugent.be/blink/editor/! I tested it with one of my projects and, once stopped, I can go back and forth. Incredibly incredibly useful.

So, why not to embed it in the ScratchAddons debugger? The source code is available at https://github.com/scratch-ed/blink

**Reply by WorldLanguages** [reply]
Most likely, it's not straight forward. Implementing a new feature by forking Scratch vs. running an addon on top of vanilla Scratch are quite different.

**Reply by sfederici** [reply]
Yes, I understand. But I guess that having a look at it could be worth a shot

**Reply by sfederici** [reply]
An explanation (I don't know how deep) is available at https://www.softxjournal.com/article/S2352-7110(23)00313-8/pdf
